adapters/pubmatic.go

```diff
@@ -44,38 +49,103 @@ type pubmaticParams struct {
 	AdSlot      string `json:"adSlot"`
 }
 
+func PrepareLogMessage(tID, pubId, adUnitId, bidID, details string, args ...interface{}) string {
+	return fmt.Sprintf("[PUBMATIC] ReqID [%s] PubID [%s] AdUnit [%s] BidID [%s] %s \n",
+		tID, pubId, adUnitId, bidID, details)
+}
+
 func (a *PubmaticAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *pbs.PBSBidder) (pbs.PBSBidSlice, error) {
 	mediaTypes := []pbs.MediaType{pbs.MEDIA_TYPE_BANNER, pbs.MEDIA_TYPE_VIDEO}
 	pbReq, err := makeOpenRTBGeneric(req, bidder, a.FamilyName(), mediaTypes, true)
 
 	if err != nil {
+		glog.Warningf("[PUBMATIC] Failed to make ortb request for request id [%s] \n", pbReq.ID)
 		return nil, err
 	}
 
+	adSlotFlag := false
+	pubId := ""
+	if len(bidder.AdUnits) > MAX_IMPRESSIONS_PUBMATIC {
+		glog.Warningf("[PUBMATIC] First %d impressions will be considered from request tid %s\n",
+			MAX_IMPRESSIONS_PUBMATIC, pbReq.ID)
+	}
+
 	for i, unit := range bidder.AdUnits {
 		var params pubmaticParams
 		err := json.Unmarshal(unit.Params, &params)
 		if err != nil {
-			return nil, err
+			glog.Warningf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
+				fmt.Sprintf("Ignored bid: invalid JSON  [%s] err [%s]", unit.Params, err.Error())))
+			continue
 		}
+
 		if params.PublisherId == "" {
-			return nil, errors.New("Missing publisherId param")
+			glog.Warningf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
+				fmt.Sprintf("Ignored bid: Publisher Id missing")))
+			continue
 		}
+		pubId = params.PublisherId
 		if params.AdSlot == "" {
-			return nil, errors.New("Missing adSlot param")
+			glog.Warningf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
+				fmt.Sprintf("Ignored bid: adSlot missing")))
+			continue
+		}
+
+		adSlotStr := strings.TrimSpace(params.AdSlot)
+		adSlot := strings.Split(adSlotStr, "@")
+		if len(adSlot) == 2 && adSlot[0] != "" && adSlot[1] != "" {
+
+			if pbReq.Imp[i].Banner != nil {
+				pbReq.Imp[i].Banner.Format = nil // pubmatic doesn't support
+				adSize := strings.Split(strings.ToLower(strings.TrimSpace(adSlot[1])), "x")
+				if len(adSize) == 2 {
+					width, err := strconv.Atoi(strings.TrimSpace(adSize[0]))
+					if err != nil {
+						glog.Warningf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
+							fmt.Sprintf("Ignored bid: invalid adSlot width [%s]", adSize[0])))
+						continue
+					}
+
+					heightStr := strings.Split(strings.TrimSpace(adSize[1]), ":")
+					height, err := strconv.Atoi(strings.TrimSpace(heightStr[0]))
+					if err != nil {
+						glog.Warningf(PrepareLogMessage(pbReq.ID, params.PublisherId, unit.Code, unit.BidID,
+							fmt.Sprintf("Ignored bid: invalid adSlot height [%s]", heightStr[0])))
+						continue
+					}
+
+					pbReq.Imp[i].TagID = strings.TrimSpace(adSlot[0])
+					pbReq.Imp[i].Banner.H = uint64(height)
+					pbReq.Imp[i].Banner.W = uint64(width)
+					adSlotFlag = true
```

Comment:
Little confused by this.

`adSlotFlag` starts out as `false`. This is the only place that it may get set to `true`. Later, you write:

```
	if !(adSlotFlag) {
		return nil, errors.New("Incorrect adSlot / Publisher param")
	}
```

Can't this whole thing be simplified by replacing this line with `return nil, errors.New("Incorrect adSlot / Publisher param")`?

Reply:
Initially ```adSlotFlag```  is set to ```false```, when valid impression is parsed it is set to ```true```. The check
``` if !(adSlotFlag)``` is added to validate valid impression is present in request or not.
And error details ```Incorrect adSlot / Publisher param``` are added to get details about invalid request.

Reply:
oh man... yeah, ignore me, i gotcha ^^